### PR TITLE
Serialize (destructure) implementation, culture aware rendering, rename destructure to Serialize

### PR DESCRIPTION
**Pull request by 304NotModified**
- Serialize (destructure) implementation, 
- Injecting Serializers
- culture aware rendering
- moved rendering methods out of template.cs
- various project.json improvements
- renamed destructure to Serialize

### CODE REVIEW

**Comment by 304NotModified** [comment]
@jods4 your opinion please ! 

NB: There is still some work to be done/todos in the code

File: src/Parser/Serialization/ISerialisation.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Text;
+
+namespace Parser
+{
+    public interface ISerialisation
```
Comment (jods4) [inline src/Parser/Serialization/ISerialisation.cs]: `ISerialization` would be more consistent with method name (or vice-versa, but .NET uses `z`).

File: src/Parser/Serialization/FuncSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Text;
+
+namespace Parser
+{
+    public class FuncSerializer<T> : ISerialisation
+        where T : class //todo add also class for struct
```
Comment (jods4) [inline src/Parser/Serialization/FuncSerializer.cs]: Can't we remove the `class` restriction (hence covering `struct`) by using a `is` + cast combo rather than `as` in the `SerializeObject` code?

File: src/Parser/Serialization/FuncSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        {
+            _func = func;
+        }
+
+        public void SerializeObject(StringBuilder sb, object value, IFormatProvider formatProvider)
+        {
+            var o = value as T;
+            if (o != null)
+            {
+                sb.Append(_func(o));
```
Comment (jods4) [inline src/Parser/Serialization/FuncSerializer.cs]: The fact that the `formatProvider` is not used (I know it comes from the interface) kind of looks like an unexpected behavior to me.
Suggested implementation for this method:
```csharp
if (o is T)
  sb.Append(Convert.ToString(o, formatProvider));
```

File: src/Parser/Serialization/DefaultSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+
+namespace Parser
+{
+
+    ///<summary>TODO combine with NLog's Json (injected) serializer?</summary>
```
Comment (jods4) [inline src/Parser/Serialization/DefaultSerializer.cs]: Yeah, I think it would be much better to use an existing JSON serializer:
- You have to write _tons_ of test cases to be sure you have a full, correct JSON implementation.
- Making it _fast_ is difficult and a lot of work.
- Then there's a ton of things you might want to support (e.g. look at Newtonsoft Json.NET): support for dictionaries, datatables, ignoring non-serializable members, etc. etc.
- And making it customizable is even worse. People invariably want lots of tweaks to naming, mappings, special types, omitting default values, etc. etc. Again just have a look at Json.NET, it's crazy.

Bottom line: I think logger and serializer should be two separated libs.
Even if you have ultra-basic internal JSON usage and don't need any fuss at all I would suggest taking a dependency on something like SimpleJSON.NET (probably not the case here).

File: src/Parser/Serialization/FuncSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Text;
+
+namespace Parser
+{
+    public class FuncSerializer<T> : ISerialisation
```
Comment (jods4) [inline src/Parser/Serialization/FuncSerializer.cs]: According to my previous comment about using a JSON lib, this wouldn't be needed? E.g. people using JSON.NET have all the extension hooks they need to customize their output and special types handling.

File: src/Parser/Serialization/SerializationManager.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+
+namespace Parser
+{
+    ///<summary>Manager serializer for customizing serialisation of specific types</summary>
+    public class SerializationManager
```
Comment (jods4) [inline src/Parser/Serialization/SerializationManager.cs:9]: Again, this would be handled by a JSON 3rd party lib, right?

File: src/Parser/TemplateRenderer.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                    else
+                    {
+                        RenderHole(sb, template.Holes[holeIndex], formatProvider, parameters[holeIndex++]);
+                    }
+                }
+            }
+            return sb.ToString();
+        }
+
+        private static void RenderHolePositional(StringBuilder sb, Hole hole, IFormatProvider formatProvider, object[] parameters)
```
Comment (jods4) [inline src/Parser/TemplateRenderer.cs]: Nitpicking but I'm not sure this method adds a lot, as the code could be directly above.
In particular, it contains as much logic as the other if-branch, which did not receive its own `RenderHoleNamed` method.

File: src/Parser/ValueRenderer.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+
+namespace Parser
+{
+    public static class ValueRenderer
```
Comment (jods4) [inline src/Parser/ValueRenderer.cs:10]: Again: this could be done by the JSON library, right?

**Review by jods4 — COMMENTED** [review]
Looks good to me, except I strongly believe that you shouldn't do the JSON serialization yourself. This is too large to implement inside a logging lib and will surely lead to tons of feature requests by users.

**Comment by jods4** [comment]
As an aside: you are probably aware but MS is adding lots of cool primitives to .NET to help with perf of code such as this one. Notably `System.Slices` to ease the parsing code, `System.Text.Formatting` for efficient, well, formatting and `System.Text.Primitives` for efficient integral parsing :)

File: src/Parser/Serialization/DefaultSerializer.cs (hunk 4 above)
Comment (304NotModified) [reply src/Parser/Serialization/DefaultSerializer.cs]: agree, see https://github.com/NLog/yamtp/issues/30 :)

File: src/Parser/ValueRenderer.cs (hunk 8 above)
Comment (304NotModified) [reply src/Parser/ValueRenderer.cs:10]: indeed!

File: src/Parser/TemplateRenderer.cs (hunk 7 above)
Comment (304NotModified) [reply src/Parser/TemplateRenderer.cs]: Agree. It was already there, so left it.

**Comment by 304NotModified** [comment]
Thanks @jods4 !!

File: src/Parser/Serialization/ISerialisation.cs (hunk 1 above)
Comment (304NotModified) [reply src/Parser/Serialization/ISerialisation.cs]: done

File: src/Parser/Serialization/FuncSerializer.cs (hunk 5 above)
Comment (304NotModified) [reply src/Parser/Serialization/FuncSerializer.cs]: this is for partials serialization. Not sure if it's there in e.g. JSON.NET (without attributes)

File: src/Parser/Serialization/FuncSerializer.cs (hunk 2 above)
Comment (304NotModified) [reply src/Parser/Serialization/FuncSerializer.cs]: done (less efficient, but we could split it later)

File: src/Parser/Serialization/FuncSerializer.cs (hunk 3 above)
Comment (304NotModified) [reply src/Parser/Serialization/FuncSerializer.cs]: this is a mistake. will pass formatProvider

File: src/Parser/Serialization/SerializationManager.cs (hunk 6 above)
Comment (304NotModified) [reply src/Parser/Serialization/SerializationManager.cs:9]: keeping this for now

File: src/Parser/TemplateRenderer.cs (hunk 7 above)
Comment (304NotModified) [reply src/Parser/TemplateRenderer.cs]: Fixed this :)

**Comment by 304NotModified** [comment]
close/reopen to trigger AppVeyor

**Comment by 304NotModified** [comment]
merged this and will continue in another branch
